src/Symfony/Bridge/Twig/Extension/AssetExtension.php

```diff
@@ -0,0 +1,136 @@
+            new \Twig_SimpleFunction('asset_version', array($this, 'getAssetVersion')),
+            new \Twig_SimpleFunction('assets_version', array($this, 'getAssetsVersion')),
+        );
+    }
+
+    /**
+     * Returns the public url/path of an asset.
+     *
+     * If the package used to generate the path is an instance of
+     * UrlPackage, you will always get a URL and not a path.
+     *
+     * @param string $path        A public path
+     * @param string $packageName The name of the asset package to use
+     *
+     * @return string The public path of the asset
+     */
+    public function getAssetUrl($path, $packageName = null, $absolute = false, $version = null)
+    {
+        // BC layer to be removed in 3.0
+        if (2 < $count = func_num_args()) {
+            trigger_error('Generating absolute URL with the Twig asset() function was deprecated in 2.7 and will be removed in 3.0. Please use absolute_url() instead.', E_USER_DEPRECATED);
+            if (4 === $count) {
+                trigger_error('Forcing a version with the Twig asset() function was deprecated in 2.7 and will be removed in 3.0.', E_USER_DEPRECATED);
+            }
+
+            $args = func_get_args();
+
+            return $this->getLegacyAssetUrl($path, $packageName, $args[2], isset($args[3]) ? $args[3] : null);
+        }
+
+        return $this->packages->getUrl($path, $packageName);
+    }
+
+    /**
+     * Returns the version of an asset.
+     *
+     * @param string $path        A public path
+     * @param string $packageName The name of the asset package to use
+     *
+     * @return string The asset version
+     */
+    public function getAssetVersion($path, $packageName = null)
+    {
+        return $this->packages->getVersion($path, $packageName);
+    }
+
+    public function getAssetsVersion($packageName = null)
+    {
+        trigger_error('The Twig assets_version() function was deprecated in 2.7 and will be removed in 3.0. Please use asset_version() instead.', E_USER_DEPRECATED);
+
+        return $this->packages->getVersion('/', $packageName);
+    }
+
+    private function getLegacyAssetUrl($path, $packageName = null, $absolute = false, $version = null)
+    {
+        if ($version) {
+            $package = $this->packages->getPackage($packageName);
+
+            $v = new \ReflectionProperty($package, 'versionStrategy');
+            $v->setAccessible(true);
+
+            $currentVersionStrategy = $v->getValue($package);
+
+            $f = new \ReflectionProperty($currentVersionStrategy, 'format');
+            $f->setAccessible(true);
+            $format = $f->getValue($currentVersionStrategy);
+
+            $v->setValue($package, new StaticVersionStrategy($version, $format));
+        }
+
+        $url = $this->packages->getUrl($path, $packageName);
+
+        if ($version) {
+            $v->setValue($package, $currentVersionStrategy);
+        }
```

Comment:
what if getting the url of the package throws an exception for some reason ? Resetting the internal state needs to be done in this case too, to avoid leaving it in a bad state